docker_registry.py

```diff
@@ -0,0 +1,212 @@
+                    reference = basename
+
+            if reference is None:
+                raise Exception("No reference determined")
+
+            return self.uploadManifest(content, reference)
+
+
+    def getManifest(self, reference):
+        """Get a (json-parsed) manifest with the given reference (digest or tag).
+        If the manifest does not exist, return None. For other errors, False."""
+        resp = self.doHttpCall("GET", "/v2/%s/manifests/%s" % (self.repository, reference),
+                               headers={'Accept': "application/vnd.docker.distribution.manifest.list.v2+json,application/vnd.docker.distribution.manifest.v2+json"})
+
+        if resp.status_code == 404:
+            return None
+
+        if resp.status_code != 200:
+            return False
+
+        return resp.json()
+
+    def getManifestDigest(self, reference):
+        """Return the digest of the manifest with the given reference.
+        If the manifest doesn't exist or the request fails, it returns False."""
+        resp = self.doHttpCall("HEAD", "/v2/%s/manifests/%s" % (self.repository, reference),
+                               headers={'Accept': "application/vnd.docker.distribution.manifest.list.v2+json,application/vnd.docker.distribution.manifest.v2+json"})
+
+        if resp.status_code != 200:
+            return False
+
+        return resp.headers['Docker-Content-Digest']
+
+    def deleteManifest(self, digest):
+        """Delete the manifest with the given reference."""
+        resp = self.doHttpCall("DELETE", "/v2/%s/manifests/%s" % (self.repository, digest))
+
+        return resp.status_code == 202
+
+    def uploadBlob(self, filename):
+        """Upload the blob with the given filename (has to equal path + digest).
+        Returns True if blob already exists or upload succeeded."""
+        digest = os.path.basename(filename)
+        if not digest.startswith("sha256:"):
+            raise Exception("Invalid filename")
+
+        # Check whether the blob already exists - don't upload it needlessly.
+        stat_request = self.doHttpCall("HEAD", "/v2/%s/blobs/%s" % (self.repository, digest))
+        if stat_request.status_code == 200 or stat_request.status_code == 307:
+            return True
+
+        # For now we can do a single upload call with everything inlined
+        # (which also means completely in ram, but currently it's never > 50 MiB)
+        content = None
+        with open(filename, "rb") as blob:
+            content = blob.read()
+
+        # First request an upload "slot", we get an URL we can PUT to back
+        upload_request = self.doHttpCall("POST", "/v2/%s/blobs/uploads/" % self.repository)
+        if upload_request.status_code == 202:
+            location = upload_request.headers["Location"]
+            upload = self.doHttpCall("PUT", location + "&digest=" + digest,
+                                     headers={'Content-Length': str(len(content))},
```

Comment:
This really has to be done manually? Not overly familiar with python's `requests`, but that seems rather basic.

Reply:
I just put it there to be safe - apparently it's not necessary:

`Content-Length headers will be overridden when we can determine the length of the content.`